dash/_validate.py

```diff
@@ -2,73 +2,76 @@
 import re
 
 from .development.base_component import Component
-from .dependencies import Input, Output, State
 from . import exceptions
 from ._utils import patch_collections_abc, _strings, stringify_id
 
 
-def validate_callback(output, inputs, state):
+def validate_callback(output, inputs, state, extra_args, types):
+    Input, Output, State = types
+    if extra_args:
+        if not isinstance(extra_args[0], (Output, Input, State)):
+            raise exceptions.IncorrectTypeException(
+                """
+                Callback arguments must be `Output`, `Input`, or `State` objects,
+                optionally wrapped in a list or tuple. We found (possibly after
+                unwrapping a list or tuple):
+                {}
+                """.format(
+                    repr(extra_args[0])
+                )
+            )
+
+        raise exceptions.IncorrectTypeException(
+            """
+            In a callback definition, you must provide all Outputs first,
+            "then all Inputs, then all States. Found this out of order:
+            {}
+            """.format(
+                repr(extra_args)
```

Comment:
When running against
```
@app.callback(
    Output('table', 'data'),
    State('table', 'data_previous'),
    Input('btn', 'n_clicks'),
    State('table', 'data')
)
```

Got the error message:
```
dash.exceptions.IncorrectTypeException: In a callback definition, you must provide all Outputs first,
"then all Inputs, then all States. Found this out of order:
[<Input `btn.n_clicks`>, <State `table.data`>]
```

As the error causing props are from the previous `State('table', 'data_previous')` and `Input('btn', 'n_clicks')`, the message could be tweaked to be more helpful / point to the problematic parameters / relationship more precisely.

Reply:
Good call. In https://github.com/plotly/dash/pull/1180/commits/d158ef8bfb4128f23578ab42d20e17d9745e9d87 I changed it so this particular situation would say:

```
dash.exceptions.IncorrectTypeException: In a callback definition, you must provide all Outputs first,
then all Inputs, then all States. After this item:
<State `table.data_previous`>
we found this item next:
<Input `btn.n_clicks`>
```